### ISSUE TITLE
Automated rollback using Argocd is it possible?

### ISSUE THREAD

**Issue by nandeeshb09** [argoproj/argo-cd#6147]
Hi Team,

Im trying to find or how to implement automated rollback using Argocd.

But found some of the options like argo rollouts, but argo rollout is not automated rollback option.

As I need to rollback based on the below conditions.

1 .Based on liveness and readiness probs
If liveness or readiness failed argocd should rollback to the last successful revision.
2 . Based on some dependency.
During the deployment, if any releases failed , then it should rollback to last successful revision

Is it available using Argocd and if its there , then how it can be achieved?

**Comment by servo1x** [comment]
I think you may want to use something like [Argo Rollouts](https://argoproj.github.io/argo-rollouts/).

**Comment by nandeeshb09** [comment]
@servo1x ,

Argo rollout is different , its rolling update to the newer version with blue green strategy or canary deployment.

Im looking for automatic rollback if any issue with current version/release.

**Comment by AlexDHoffer** [comment]
I agree with @nandeeshb09. I'm a bit confused why there's not an option built in just to roll back to the last commit if an app is degraded. I don't see why you need a whole separate other tool just for something simple like what Helm natively provides with ``helm install --atomic``.

**Comment by musabmasood** [comment]
May be its easier to implement without `autosync` enabled, but with `autosync` in place it becomes even trickier. 

A "Healthy" application when "autosync" is enabled implies that the deployed version is the same as the branch HEAD / latest commit. If the deployment fails and it points to the last commit, the implication does't hold true anymore. 

May be there could be another status like "Rolled back" but then what if the rollback fails as well?

**Comment by jackmtrayport** [comment]
Is there any plan to implement something like this? We don't plan on using automatic sync, so a simple option that says "if the app failed to sync, automatically roll back to the previously-synced version and sync that" would suffice.

**Comment by youssefBa1** [comment]
hey did find a way to achieve an Automated rollback

**Comment by Tanmoy037** [comment]
Hi, this feature is useful. Need this feature! Any plans for this one?

**Comment by HujinoKun** [comment]
Hello, 

Any news on this topic?

**Comment by gaeljw** [comment]
ArgoCD is meant to be used with auto-sync. That is: state in git = state in the cluster, no manual intervention.

Automatic rollback goes a bit against that. It would require ArgoCD to commit back to Git. But it's not ArgoCD that should be responsible to define the state in Git. How would one even notice that the last commit was rollbacked and the desired version is not deployed?

Even if the event auto-sync is disabled, when/how will ArgoCD have a chance to retry to sync? It's one of its first role to keep trying to sync the desired state.

Also, if your pod fails the probes, it's standard K8S Deployment strategy to stop at the 1st pod and not continue. Isn't that enough? This has the benefit to also raise alerts automatically as you've got both the ArgoCD app in not healthy state + a Pod that keeps crashing, your monitoring/alerting should tell you. And from there you can automate some actions.

To be clear: I'm not a maintainer but I don't think that fits in the recommended usage of ArgoCD in the 1st place.

Argo Rollouts with custom metrics can maybe handle this as another comment said.

**Comment by jackmtpt** [comment]
>It would require ArgoCD to commit back to Git

Would it? Imagine that the sync had failed for whatever reason (maybe ArgoCD lost connection to the apiserver, doesn't matter really) - ArgoCD knows that the cluster doesn't match git, it's tried to reconcile it and wasn't able to. If you don't have auto-sync enabled then it just stops there, with the Application resource in a failed state. Nothing gets written back to git here, and I don't think an automatic rollback would need to either; ArgoCD can sync the previously-synced commit again, mark the Application as failed and then stop.

Tbh we don't use auto-sync for anything because we're not allowed to, so I'm not sure what the current behaviour on a failed auto-sync would be, but presumably anyone who cares what the state of their applications is already had monitoring in place to detect sync failures so even if ArgoCD stops trying to sync the failed commit that would be fine IMO.

>Also, if your pod fails the probes, it's standard K8S Deployment strategy to stop at the 1st pod and not continue. Isn't that enough?

Not for stateful applications where you need to get a shard/partition of something back online.

>Argo Rollouts with custom metrics can maybe handle this as another comment said.

In practice this is exactly what we're doing, and we're not using ArgoCD for our production applications. A pipeline runs `helm upgrade` which kicks off a Rollout, and ArgoCD is limited to infrastructure type applications only.

**Comment by gaeljw** [comment]
For the record, looking a bit more at Argo Rollouts, here's a great explanation on how they handle rollbacks: https://argo-rollouts.readthedocs.io/en/stable/FAQ/#if-i-use-both-argo-rollouts-and-argo-cd-wouldnt-i-have-an-endless-loop-in-the-case-of-a-rollback

**Comment by crenshaw-dev** [comment]
The devil is very much in the details with a proposal like this.

To do automatic rollouts, the system needs to know: 1) is the Application in a Bad state, 2) what was the most recent Good state, and 3) is it safe to return to that previous state?

Argo CD health checks are very rough approximations of an Application's actual health. They're meant for a human to quickly understand the high-level state of their resources. They're not a substitute for proper, use-case-specific monitoring. So we should be immediately suspicious of Argo CD's idea of "Degraded" and "last Healthy" states.

Even if we could 100% trust those values, re-applying the old state isn't necessarily safe. We had a situation at Intuit last week where a service's app code made a DB schema change. The old code wasn't compatible with the new schema and started throwing 5xx errors. Even though the old Pods were "Degraded", reverting to the "last Healthy state" would be the worst possible decision, because it would have sent all traffic to the broken code.

Argo Rollouts helps ensure changes are backwards-compatible by running both versions of the code at once. Developers quickly realize that if they deploy backwards-incompatible code, things will break. But arbitrary Kubernetes resources don't provide this kind of protection.

[Hazel Weakly](https://hazelweakly.me/) recently posted an interesting idea on LinkedIn: always roll back as part of every deployment. This not only validates the rollback process, it helps teach developers that _all_ changes (including external side-effects) must be backwards compatible. If developers don't follow this rule, they quickly receive feedback of their mistake.

tl;dr: Argo CD is poorly positioned to do automated rollbacks well. If someone feels their Application qualifies for automated rollbacks, they can implement it in CI or in a use-case-specific operator.

Closing to express that this feature is not currently planned but **not to foreclose discussion**. I'd love if we continue to discuss the particulars of automated rollbacks here!
